Implements back the port based SSL certificate fetching + real SNI support.

GitHubProUser67 opened the pull request:
Implements back the port based SSL certificate fetching.

This modernize the Mono code base to allow booth, a in-memory cache and that same file-based cache if the in-memory cache doesn't contain the entry.

This makes the previous SetCertificate constructor obsolete, but for good, since we can now set a different certificate per ports.

This is not SNI level, but it's still leagues beyond the actual .NET implementation (the MYSTICAL null stubbed return value with the TODO ^^).

GitHubProUser67 committed:
Implements back the port based SSL certificate fetching.

Implements back the port based SSL certificate fetching.

This modernize the Mono code base to allow booth, a in-memory cache and that same file-based cache if the in-memory cache doesn't contain the entry.

This makes the previous SetCertificate constructor obsolete, but for good, since we can now set a different certificate per ports.

PJB3005 commented:
Do you have an actual use case for this? The reason the mono code even *has* that port-based logic is dated and ancient, I don't see any reason to keep it as functionality here.

GitHubProUser67 commented:
I do have an use case if I host various CN names on various ports without having to rely on SAN attributes.

PJB3005 commented:
> I do have an use case if I host various CN names on various ports without having to rely on SAN attributes.

Do *you* have an actual use case for this? Sure obviously this would allow you to have separate TLS certs served per port but *what* use case could that possibly have in 2024 to warrant implementing?

I would rather have a more flexible API that also incorporated SNI support or similar via e.g. a callback. Then you can just add the destination socket port in one of the parameters the selection logic and users that *really* need this can take it from there.

GitHubProUser67 committed:
Implements SNI Support.

Implements SNI Support.

Uses the same constructor as the previous solution, instead you will need to pass-in the CA Certificate.

GitHubProUser67 committed:
HttpConnection - fixes non-https path.

HttpConnection - fixes non-https path.

GitHubProUser67 committed:
HttpListener.Certificates - Implement certificate Authority input validation for older NET SDKs.

HttpListener.Certificates - Implement certificate Authority input validation for older NET SDKs.

PJB3005 commented:
> and HTTPListener will sign itself intermediate certificates that are instanced per-session.

Uh, what? Does that actually work? You can't self-sign certificates. And you can't get a root CA cert from anywhere? And isn't signing way too expensive to do for every TLS connection?

GitHubProUser67 commented:
I can confirm it works just fine, the root CA will need to be obtained from an outside source (I can make a generator for it too if needed), then this CA cert will be used to sign all the certificates that are generated on the fly.

They are all valid for a period of 7 days to avoids making too much of them (this can be extended).

GitHubProUser67 committed:
Prevents assertion failed crash with an empty url.

Some web-crawlers tries to send a request with an empty url, we forbid these in order to not throw a debug assertion.

PJB3005 commented on src/System/Net/CertificateHelper.cs:
```diff
@@ -0,0 +1,233 @@
+{
+    public class CertificateHelper
+    {
+        private static readonly string[] tlds = {
```
This is an extremely outdated list of TLDs. You cannot hardcode something like this.

PJB3005 commented on src/System/Net/CertificateHelper.cs:
This organization does not exist and it should not be hardcoded in the library.

PJB3005 commented on src/System/Net/CertificateHelper.cs:
This email address does not exist and it should not be hardcoded in the library.

PJB3005 commented on src/System/Net/CertificateHelper.cs:
This rando isn't cryptographically secure.

PJB3005 commented on src/System/Net/CertificateHelper.cs:
This random isn't cryptographically secure.

PJB3005 commented on src/System/Net/CertificateHelper.cs:
This is not a correct way to check this, nor should this library care.

PJB3005 commented on src/System/Net/CertificateHelper.cs:
We are not implementing workarounds to security issues like this.

PJB3005 commented on src/System/Net/CertificateHelper.cs:
.NET has built-in libraries for this.

PJB3005 commented:
I am closing this PR. Based on the code I have skimmed over so far it is clear to me that:

1. This code has absolutely no place, whatsoever, in being in a library like this.
2. I hate to say this, but you likely do not know what you are doing when it comes to TLS certificates.

If you want to have a way to specify custom certificates dynamically like this, a PR that adds a simple callback so that library consumers can pass their own cert is all that should be necessary. You can implement whatever... this is.. on top of that.

GitHubProUser67 commented:
Okay, most of these are valid points I admit.

However, if I am being completely honest, I do not understand, nor like, the 2. sentence there, I would like some examples and proof as to where I didn't understand how TLS worked.

Some code might not be the best way to put it, but, as far as I know, the certs "works" so does TLS, it is simply skimming over some security, which is a valid point for it not being merged.

PJB3005 commented:
> However, if I am being completely honest, I do not understand, nor like, the 2. sentence there, I would like some examples and proof as to where I didn't understand how TLS worked.

You are self-signing certificates to implement SNI, directly for any domain clients ask for. This is not something you can do in the real world unless you have your own CA certs preinstalled on client machines. Even if you did have that, it would be grossly irresponsible from a security perspective to have your web server self signing certificates right then and there for anything other than dev environment testing.

metalgearsloth commented:
> as far as I know, the certs "works" so does TLS, it is simply skimming over some security,

Isn't security the whole point of the exercise?

GitHubProUser67 commented:
I better understand the point now, the security argument is valid, I did use this system so old systems where SSL is a requirement could work.

On my fork I re-did that a bit, to also support chain signed certificate from an other CA (by loading the file directly if it exists) with a fallback to that system if not found.

But after all, this is quite use-case specific, so maybe it is best suited for the fork only.
